src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -260,6 +289,224 @@ public Subscription getSubscription() {
+        if (c == null) {
+            throw new IllegalStateException("Message is not bound to a connection");
+        }
+        return c;
+
+    }
+
+    private void ackReply(byte[] ackType) {
+        try {
+            ackReply(ackType, Duration.ZERO);
+        } catch (InterruptedException e) {
+            // we should never get here, but satisfy the linters.
+            Thread.currentThread().interrupt();
+        } catch (TimeoutException e) {
+            // NOOP
+        }
+    }
+    
+    private boolean isPullMode() {
+        if (!(this.subscription instanceof NatsJetStreamSubscription)) {
+            return false;
+        }
+        return (((NatsJetStreamSubscription) this.subscription).pull > 0);
+    }
+
+    private void ackReply(byte[] ackType, Duration d) throws InterruptedException, TimeoutException {
+        if (!this.isJetStream()) {
+            throw new IllegalStateException("Message is not a jestream message");
+        }
+        if (d == null) {
+            throw new IllegalArgumentException("Duration cannot be null.");
+        }
+
+        boolean isSync = (d != Duration.ZERO);
+        Connection nc = getJetStreamValidatedConnection();
+
+        if (isPullMode()) {
+           if (Arrays.equals(ackType, AckAck)) {
+              nc.publish(replyTo, subscription.getSubject(), AckNext);
+           } else if (Arrays.equals(ackType, AckNak) || Arrays.equals(ackType, AckTerm)) {
+                nc.publish(replyTo, subscription.getSubject(), AckNextOne);
+           }
+           if (isSync && nc.request(replyTo, null, d) == null) {
+                throw new TimeoutException("Ack request next timed out.");
+           }
+
+        } else if (isSync && nc.request(replyTo, ackType, d) == null) {
+            throw new TimeoutException("Ack response timed out.");
+        } else {
+            nc.publish(replyTo, ackType);
+        }
+    }
+
+    @Override
+    public void ack() {
+        ackReply(AckAck);
+    }
+
+    @Override
+    public void ackSync(Duration d) throws InterruptedException, TimeoutException {
+        ackReply(AckAck, d);
+    }
+
+    @Override
+    public void nak(){
+        ackReply(AckNak);
+    }
+
+    @Override
+    public void inProgress() {
+        ackReply(AckProgress);
+    }
+
+    @Override
+    public void term() {
+        ackReply(AckTerm);
+    }
+
+    public void toDOackNextRequest(ZonedDateTime expiry, long batch, boolean noWait) {
+
+        if (batch < 0) {
+            throw new IllegalArgumentException();
+        }
+
+        Connection c = getConnection();
+        if (c == null) {
+            throw new IllegalStateException("Message is not bound to a connection");
+        }
+        
+        // minor optimization for the ack.
+        byte[] payload;
+        if (expiry == null && batch == 0 && !noWait) {
+            payload = AckNextEmptyPayload;
+        } else {
+            StringBuilder sb = new StringBuilder("+ACKNXT {");

```

Comment:
good candidate for ByteArrayBuilder

Reply:
Yeah...  I ended up refactoring this code entirely so there's no string builder anymore here.